README.md
# circle-ci-config-linter

[![CircleCI](https://circleci.com/gh/20k-ultra/circle-ci-config-linter.svg?style=shield)](https://circleci.com/gh/20k-ultra/circle-ci-config-linter)
[![CircleCI-Linter](https://qnld475cd3.execute-api.us-east-2.amazonaws.com/beta?repo=20k-ultra/circle-ci-config-linter)](https://github.com/20k-ultra/circle-ci-config-linter) <-- Generated by this tool!

Add a lint errors badge on your repos [CircleCI](https://circleci.com/) config. You can learn more about CircleCI config options at [docs](https://circleci.com/docs/2.0/configuration-reference/).

Replace {{YOUR-REPOSITORY}} to your profile or org and repository name.

```
[![CircleCI-Linter](https://qnld475cd3.execute-api.us-east-2.amazonaws.com/beta?repo={{YOUR-REPOSITORY}})](https://github.com/20k-ultra/circle-ci-config-linter)
```

### Examples

[CircleCI-Public / circleci-cli](https://github.com/CircleCI-Public/circleci-cli/blob/master/.circleci/config.yml)

[![CircleCI-Linter](https://qnld475cd3.execute-api.us-east-2.amazonaws.com/beta?repo=CircleCI-Public/circleci-cli)](https://github.com/20k-ultra/circle-ci-config-linter)

[juxt / jinx](https://github.com/juxt/jinx//blob/master/.circleci/config.yml)

[![CircleCI-Linter](https://qnld475cd3.execute-api.us-east-2.amazonaws.com/beta?repo=juxt/jinx)](https://github.com/20k-ultra/circle-ci-config-linterl)

Repo without CircleCI config

[![CircleCI-Linter](https://qnld475cd3.execute-api.us-east-2.amazonaws.com/beta?repo=DNSCrypt/dnscrypt-proxy)](https://github.com/20k-ultra/circle-ci-config-linter)

### Building

```bash
$ lein install
$ lein test
```

### To do

- allow for passing parameter so that details about errors are returned instead of an SVG.
- add more button styles (See [shields.io](https://shields.io))
- add CircleCI AWS [CLI](https://circleci.com/orbs/registry/orb/circleci/aws-cli) or [Serverless](https://circleci.com/orbs/registry/orb/circleci/aws-sam-serverless) Orbs to automatically deploy builds

### Mentions

- [Schemastore](https://github.com/SchemaStore/schemastore/)
- [json-schema.clj](https://github.com/niquola/json-schema.clj)

The Schemastore spec used for this can be found in [/resources/circle-ci-schema.json](https://github.com/20k-ultra/circle-ci-config-linter/blob/master/resources/circle-ci-schema.json). I had to modify it slightly because the original was causing json-schema.clj to throw some errors.


## License

Copyright © 2020

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2020-03-01
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2020-03-01
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/circle-ci-config-linter/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/circle-ci-config-linter/compare/0.1.0...0.1.1


test/circle_ci_config_linter/core_test.clj
(ns circle-ci-config-linter.core-test
  (:require [clojure.test :refer :all]
            [clojure.java.io :as io]
            [circle-ci-config-linter.core :refer :all]))

(def test-data {:id 44 :price 2.99 :greetings ["hello" "hi" "good morning"] :country {:name "Canada"}})
(def plastic-button-with-zero-errors "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" width=\"190\" height=\"20\"> <linearGradient id=\"b\" x2=\"0\" y2=\"100%\"> <stop offset=\"0\" stop-color=\"#bbb\" stop-opacity=\".1\" /> <stop offset=\"1\" stop-opacity=\".1\" /> </linearGradient> <clipPath id=\"a\"> <rect width=\"190\" height=\"20\" rx=\"3\" fill=\"#fff\" /> </clipPath> <g clip-path=\"url(#a)\"> <path fill=\"#555\" d=\"M0 0h129v20H0z\" /> <path fill=\"#4c1\" d=\"M129 0h61v20H129z\" /> <path fill=\"url(#b)\" d=\"M0 0h190v20H0z\" /> </g> <g fill=\"#fff\" text-anchor=\"middle\" font-family=\"DejaVu Sans,Verdana,Geneva,sans-serif\" font-size=\"110\"> <text x=\"655\" y=\"150\" fill=\"#010101\" fill-opacity=\".3\" transform=\"scale(.1)\" textLength=\"1190\">circle-ci-config-lint</text> <text x=\"655\" y=\"140\" transform=\"scale(.1)\" textLength=\"1190\">circle-ci-config-lint</text> <text x=\"1588\" y=\"150\" fill=\"#010101\" fill-opacity=\".3\" transform=\"scale(.1)\" textLength=\"510\"> 0 errors </text> <text x=\"1588\" y=\"140\" transform=\"scale(.1)\" textLength=\"510\"> 0 errors </text> </g> </svg>")
(def plastic-button-with-errors "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" width=\"190\" height=\"20\"> <linearGradient id=\"b\" x2=\"0\" y2=\"100%\"> <stop offset=\"0\" stop-color=\"#bbb\" stop-opacity=\".1\" /> <stop offset=\"1\" stop-opacity=\".1\" /> </linearGradient> <clipPath id=\"a\"> <rect width=\"190\" height=\"20\" rx=\"3\" fill=\"#fff\" /> </clipPath> <g clip-path=\"url(#a)\"> <path fill=\"#555\" d=\"M0 0h129v20H0z\" /> <path fill=\"#e05d44\" d=\"M129 0h61v20H129z\" /> <path fill=\"url(#b)\" d=\"M0 0h190v20H0z\" /> </g> <g fill=\"#fff\" text-anchor=\"middle\" font-family=\"DejaVu Sans,Verdana,Geneva,sans-serif\" font-size=\"110\"> <text x=\"655\" y=\"150\" fill=\"#010101\" fill-opacity=\".3\" transform=\"scale(.1)\" textLength=\"1190\">circle-ci-config-lint</text> <text x=\"655\" y=\"140\" transform=\"scale(.1)\" textLength=\"1190\">circle-ci-config-lint</text> <text x=\"1588\" y=\"150\" fill=\"#010101\" fill-opacity=\".3\" transform=\"scale(.1)\" textLength=\"510\"> 5 errors </text> <text x=\"1588\" y=\"140\" transform=\"scale(.1)\" textLength=\"510\"> 5 errors </text> </g> </svg>")
(def plastic-button-with-no-config "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" width=\"190\" height=\"20\"> <linearGradient id=\"b\" x2=\"0\" y2=\"100%\"> <stop offset=\"0\" stop-color=\"#bbb\" stop-opacity=\".1\" /> <stop offset=\"1\" stop-opacity=\".1\" /> </linearGradient> <clipPath id=\"a\"> <rect width=\"190\" height=\"20\" rx=\"3\" fill=\"#fff\" /> </clipPath> <g clip-path=\"url(#a)\"> <path fill=\"#555\" d=\"M0 0h129v20H0z\" /> <path fill=\"#e05d44\" d=\"M129 0h61v20H129z\" /> <path fill=\"url(#b)\" d=\"M0 0h190v20H0z\" /> </g> <g fill=\"#fff\" text-anchor=\"middle\" font-family=\"DejaVu Sans,Verdana,Geneva,sans-serif\" font-size=\"110\"> <text x=\"655\" y=\"150\" fill=\"#010101\" fill-opacity=\".3\" transform=\"scale(.1)\" textLength=\"1190\">circle-ci-config-lint</text> <text x=\"655\" y=\"140\" transform=\"scale(.1)\" textLength=\"1190\">circle-ci-config-lint</text> <text x=\"1588\" y=\"150\" fill=\"#010101\" fill-opacity=\".3\" transform=\"scale(.1)\" textLength=\"510\"> no config </text> <text x=\"1588\" y=\"140\" transform=\"scale(.1)\" textLength=\"510\"> no config </text> </g> </svg>")
(def plastic-button-good-config "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" width=\"190\" height=\"20\"> <linearGradient id=\"b\" x2=\"0\" y2=\"100%\"> <stop offset=\"0\" stop-color=\"#bbb\" stop-opacity=\".1\" /> <stop offset=\"1\" stop-opacity=\".1\" /> </linearGradient> <clipPath id=\"a\"> <rect width=\"190\" height=\"20\" rx=\"3\" fill=\"#fff\" /> </clipPath> <g clip-path=\"url(#a)\"> <path fill=\"#555\" d=\"M0 0h129v20H0z\" /> <path fill=\"#4c1\" d=\"M129 0h61v20H129z\" /> <path fill=\"url(#b)\" d=\"M0 0h190v20H0z\" /> </g> <g fill=\"#fff\" text-anchor=\"middle\" font-family=\"DejaVu Sans,Verdana,Geneva,sans-serif\" font-size=\"110\"> <text x=\"655\" y=\"150\" fill=\"#010101\" fill-opacity=\".3\" transform=\"scale(.1)\" textLength=\"1190\">circle-ci-config-lint</text> <text x=\"655\" y=\"140\" transform=\"scale(.1)\" textLength=\"1190\">circle-ci-config-lint</text> <text x=\"1588\" y=\"150\" fill=\"#010101\" fill-opacity=\".3\" transform=\"scale(.1)\" textLength=\"510\"> 0 errors </text> <text x=\"1588\" y=\"140\" transform=\"scale(.1)\" textLength=\"510\"> 0 errors </text> </g> </svg>")
(def plastic-button-some-errors-config "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" width=\"190\" height=\"20\"> <linearGradient id=\"b\" x2=\"0\" y2=\"100%\"> <stop offset=\"0\" stop-color=\"#bbb\" stop-opacity=\".1\" /> <stop offset=\"1\" stop-opacity=\".1\" /> </linearGradient> <clipPath id=\"a\"> <rect width=\"190\" height=\"20\" rx=\"3\" fill=\"#fff\" /> </clipPath> <g clip-path=\"url(#a)\"> <path fill=\"#555\" d=\"M0 0h129v20H0z\" /> <path fill=\"#e05d44\" d=\"M129 0h61v20H129z\" /> <path fill=\"url(#b)\" d=\"M0 0h190v20H0z\" /> </g> <g fill=\"#fff\" text-anchor=\"middle\" font-family=\"DejaVu Sans,Verdana,Geneva,sans-serif\" font-size=\"110\"> <text x=\"655\" y=\"150\" fill=\"#010101\" fill-opacity=\".3\" transform=\"scale(.1)\" textLength=\"1190\">circle-ci-config-lint</text> <text x=\"655\" y=\"140\" transform=\"scale(.1)\" textLength=\"1190\">circle-ci-config-lint</text> <text x=\"1588\" y=\"150\" fill=\"#010101\" fill-opacity=\".3\" transform=\"scale(.1)\" textLength=\"510\"> 5 errors </text> <text x=\"1588\" y=\"140\" transform=\"scale(.1)\" textLength=\"510\"> 5 errors </text> </g> </svg>")

(deftest test-read-json
  (testing
    (is (= test-data (read-json (.getPath (io/resource "test.json")))))))

(deftest test-read-yml
  (testing
    (is (= test-data (read-yml (.getPath (io/resource "test.yml")))))))

(deftest test-plastic-button-no-errors
  (testing
    (is (= plastic-button-with-zero-errors (plastic-button true "0 errors")))))

(deftest test-plastic-button-with-errors
  (testing
    (is (= plastic-button-with-errors (plastic-button false "5 errors")))))

(deftest test-plastic-button-with-no-config
  (testing
    (is (= plastic-button-with-no-config (plastic-button false "no config")))))

(deftest test-make-button-default-style
  (testing
    (is (= plastic-button-with-zero-errors (make-button true "0 errors")))))

(deftest test-make-button-default-style
  (testing
    (is (= plastic-button-with-zero-errors (make-button true "0 errors" :plastic)))))

(deftest test-linting-no-config
  (testing
    (is (= {:body plastic-button-with-no-config} (handle-event {"queryStringParameters" {}})))))

(deftest test-linting-no-error-default-style-config
  (testing
    (is (= {:body plastic-button-good-config} (handle-event {"queryStringParameters" {"repo" "juxt/jinx"}})))))

(deftest test-linting-error-default-style-config
  (testing
    (is (= {:body plastic-button-some-errors-config} (handle-event {"queryStringParameters" {"repo" "CircleCI-Public/circleci-cli"}})))))

src/circle_ci_config_linter/core.clj
(ns circle-ci-config-linter.core
  (:require [uswitch.lambada.core :refer [deflambdafn]]
            [clojure.data.json :as json]
            [clojure.java.io :as io]
            [json-schema.core :as sch]
            [cheshire.core :as chesire]
            [clj-yaml.core :as yaml]
            [clojure.core.match :refer [match]]))

(defn read-json [path] (chesire/parse-string (slurp path) keyword))
(defn read-yml [path] (yaml/parse-string (slurp path)))

(defn plastic-button [valid text-output]
  (let [button-color (if valid "#4c1" "#e05d44")]
    (str "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" width=\"190\" height=\"20\"> <linearGradient id=\"b\" x2=\"0\" y2=\"100%\"> <stop offset=\"0\" stop-color=\"#bbb\" stop-opacity=\".1\" /> <stop offset=\"1\" stop-opacity=\".1\" /> </linearGradient> <clipPath id=\"a\"> <rect width=\"190\" height=\"20\" rx=\"3\" fill=\"#fff\" /> </clipPath> <g clip-path=\"url(#a)\"> <path fill=\"#555\" d=\"M0 0h129v20H0z\" /> <path fill=\"" button-color "\" d=\"M129 0h61v20H129z\" /> <path fill=\"url(#b)\" d=\"M0 0h190v20H0z\" /> </g> <g fill=\"#fff\" text-anchor=\"middle\" font-family=\"DejaVu Sans,Verdana,Geneva,sans-serif\" font-size=\"110\"> <text x=\"655\" y=\"150\" fill=\"#010101\" fill-opacity=\".3\" transform=\"scale(.1)\" textLength=\"1190\">circle-ci-config-lint</text> <text x=\"655\" y=\"140\" transform=\"scale(.1)\" textLength=\"1190\">circle-ci-config-lint</text> <text x=\"1588\" y=\"150\" fill=\"#010101\" fill-opacity=\".3\" transform=\"scale(.1)\" textLength=\"510\"> " text-output " </text> <text x=\"1588\" y=\"140\" transform=\"scale(.1)\" textLength=\"510\"> " text-output " </text> </g> </svg>")))

(defn make-button
  ([valid text-output style]
    (match [style]
      [:plastic] (plastic-button valid text-output)
      :else (plastic-button valid text-output)))
  ([valid text-output]
    (make-button valid text-output :plastic)))

(defn handle-event
  [event]
  (def REPO ((event "queryStringParameters") "repo"))
  (def BRANCH "master")
  (try
    (def config (read-yml (str "https://raw.githubusercontent.com/" REPO "/" BRANCH "/.circleci/config.yml")))
    (catch java.io.FileNotFoundException e (def config nil))
    (catch Exception e (def config nil) (println (str "Caught: " (.toString e)))))
  (if (nil? config)
    (let [svg (make-button false "no config")]
      (hash-map :body svg))
    (let [schema (read-json (.getPath (io/resource "circle-ci-schema.json")))
        errors (:errors (sch/validate schema config))
        error-count (count errors)]
      (let [svg (make-button (= error-count 0) (str error-count " errors"))]
        (hash-map :body svg)))))

(deflambdafn circle-ci.lambda.linter
  [in out ctx]
  (let [event (json/read (io/reader in))
        res (handle-event event)]
    (with-open [w (io/writer out)]
      (json/write res w))))
